### ISSUE TITLE
zcashd crash on startup: Assertion failed: pindex->hashFinalOrchardRoot == orchardWallet.GetLatestAnchor

### ISSUE THREAD

**Issue by AArnott** [zcash/zcash#6044]
### Describe the issue

zcashd no longer gets past the loading phase.

#### If so, please list the steps to reproduce below:

run the zcashd.exe program

### Expected behaviour

Zcashd should finish loading and remain running.

### Actual behaviour + errors

zcashd fails, and doesn't even explain why in debug.log, which I include anyway below

### The version of Zcash you were using:

Zcash Daemon version v5.0.0
64-bit Windows version. Downloaded from the github releases page.

### Machine specs:
- OS name + version: Windows 11
- CPU: AMD Ryzen 9
- RAM: 64GB
- Disk size: 1.8TB (with 873GB remaining)
- Disk Type (HD/SDD): SSD
- Linux kernel version (uname -a): N/A
- Compiler version (gcc --version): N/A
- Linker version (ld -v): N/A
- Assembler version (as --version): N/A

### Any extra information that might be useful in the debugging process.

The debug.log doesn't seem to contain any hint as to the failure:

```
2022-07-02T13:52:36.050477Z  INFO main: Zcash version v5.0.0 (2022-05-10 20:14:00 -0600)
2022-07-02T13:52:36.050656Z  INFO Init: main: InitParameterInteraction: parameter interaction: -whitelistforcerelay=1 -> setting -whitelistrelay=1
2022-07-02T13:52:36.074525Z  INFO Init: main: Using BerkeleyDB version Berkeley DB 6.2.23: (March 28, 2016)
2022-07-02T13:52:36.074834Z  INFO Init: main: Default data directory C:\Users\andarno\AppData\Roaming\Zcash
2022-07-02T13:52:36.074847Z  INFO Init: main: Using data directory C:\Users\andarno\AppData\Roaming\Zcash
2022-07-02T13:52:36.075069Z  INFO Init: main: Using config file C:\Users\andarno\AppData\Roaming\Zcash\zcash.conf
2022-07-02T13:52:36.075088Z  INFO Init: main: Using at most 125 connections (2048 file descriptors available)
2022-07-02T13:52:36.075100Z  INFO Init: main: Using 16 threads for script verification
2022-07-02T13:52:36.076102Z  INFO main: scheduler thread start
2022-07-02T13:52:36.076292Z  INFO Init: main: Loading Sapling (Spend) parameters from C:\Users\andarno\AppData\Roaming\ZcashParams\sapling-spend.params
2022-07-02T13:52:36.076304Z  INFO Init: main: Loading Sapling (Output) parameters from C:\Users\andarno\AppData\Roaming\ZcashParams\sapling-output.params
2022-07-02T13:52:36.076308Z  INFO Init: main: Loading Sapling (Sprout Groth16) parameters from C:\Users\andarno\AppData\Roaming\ZcashParams\sprout-groth16.params
2022-07-02T13:52:37.048380Z  INFO Init: main: Loading Orchard parameters
2022-07-02T13:52:40.639251Z  INFO Init: main: Loaded proof system parameters in 4.563062s seconds.
2022-07-02T13:52:40.642636Z  INFO Init: main: HTTP: creating work queue of depth 16
2022-07-02T13:52:40.642814Z  INFO Init: main: Config options rpcuser and rpcpassword will soon be deprecated. Locally-run instances may remove rpcuser to use cookie-based auth, or may be replaced with rpcauth. Please see share/rpcuser for rpcauth auth generation.
2022-07-02T13:52:40.642845Z  INFO Init: main: HTTP: starting 4 worker threads
2022-07-02T13:52:40.643184Z  INFO Init: main: Using wallet wallet.dat
2022-07-02T13:52:40.643200Z  INFO Init: main: init message: Verifying wallet...
2022-07-02T13:52:40.643368Z  INFO Init: main: CDBEnv::Open: LogDir=C:\Users\andarno\AppData\Roaming\Zcash\database ErrorFile=C:\Users\andarno\AppData\Roaming\Zcash\db.log
2022-07-02T13:52:40.818100Z  INFO Init: main: Bound to [::]:8233
2022-07-02T13:52:40.818657Z  INFO Init: main: Bound to 0.0.0.0:8233
2022-07-02T13:52:40.818946Z  INFO Init: main: Cache configuration:
2022-07-02T13:52:40.818965Z  INFO Init: main: * Using 2.0MiB for block index database
2022-07-02T13:52:40.818971Z  INFO Init: main: * Using 120.0MiB for chain state database
2022-07-02T13:52:40.818974Z  INFO Init: main: * Using 328.0MiB for in-memory UTXO set
2022-07-02T13:52:40.818980Z  INFO Init: main: init message: Loading block index...
2022-07-02T13:52:40.819181Z  INFO Init: main: Opening LevelDB in C:\Users\andarno\AppData\Roaming\Zcash\blocks\index
2022-07-02T13:52:40.877771Z  INFO Init: main: Opened LevelDB successfully
2022-07-02T13:52:40.877938Z  INFO Init: main: Opening LevelDB in C:\Users\andarno\AppData\Roaming\Zcash\chainstate
2022-07-02T13:52:40.894098Z  INFO Init: main: Opened LevelDB successfully
2022-07-02T13:53:12.799962Z  INFO Init: main: LoadBlockIndexDB: last block file = 278
2022-07-02T13:53:12.800891Z  INFO Init: main: LoadBlockIndexDB: last block file info: CBlockFileInfo(blocks=2598, size=72997629, heights=1719720...1722357, time=2022-06-29...2022-07-01)
2022-07-02T13:53:12.800905Z  INFO Init: main: Checking all blk files are present...
2022-07-02T13:53:13.166657Z  INFO Init: main: LoadBlockIndexDB: transaction index disabled
2022-07-02T13:53:13.166686Z  INFO Init: main: LoadBlockIndexDB: insight explorer disabled
2022-07-02T13:53:13.166689Z  INFO Init: main: LoadBlockIndexDB: light wallet daemon disabled
2022-07-02T13:53:14.082207Z  INFO Init: main: LoadBlockIndexDB: hashBestChain=000000000187a16c268d82184175940dfbeb68a5049608cfe53f579f5bcb8331 height=1722332 date=2022-07-01 11:48:12 progress=0.999231
2022-07-02T13:53:14.085053Z  INFO Init: main: init message: Rewinding blocks if needed...
2022-07-02T13:53:19.924217Z  INFO Init: main: init message: Verifying blocks...
2022-07-02T13:53:19.924266Z  INFO Init: main: Verifying last 288 blocks at level 3
2022-07-02T13:53:22.252153Z  INFO Init: main: No coin database inconsistencies in last 289 blocks (1149 transactions)
2022-07-02T13:53:22.254012Z  INFO Init: main:  block index           41435ms
2022-07-02T13:53:22.254459Z  INFO Init: main: init message: Loading wallet...
2022-07-02T13:53:22.560756Z  INFO Init: main: nFileVersion = 5000050
2022-07-02T13:53:22.560787Z  INFO Init: main: Keys: 227 plaintext, 0 encrypted, 227 w/ metadata, 227 total
2022-07-02T13:53:22.560796Z  INFO Init: main: ZKeys: 13 plaintext, 0 encrypted, 13 w/metadata, 13 total
2022-07-02T13:53:22.599507Z  INFO Init: main:  wallet                  346ms
2022-07-02T13:53:22.599862Z  INFO Init: main: init message: Rescanning...
2022-07-02T13:53:22.599879Z  INFO Init: main: CWallet::InitLoadWallet(): Rescanning last 1 blocks (from block 1722331)...
2022-07-02T13:53:22.599903Z  INFO Init: main: CWallet::ScanForWalletTransactions(): Rewinding Orchard wallet to height 1722330; current is 1722404
```

But the console window itself *briefly* flashes an assertion failure before deleting it. I managed to catch it in an animated gif for you. It says:

> Assertion failed: pindex->hashFinalOrchardRoot == orchardWallet.GetLatestAnchor(), file wallet/wallet.cpp, line 2696

![4e479d06-9fa6-4c73-8581-1e4b0d5ddc82](https://user-images.githubusercontent.com/3548/177004147-c5a275d2-2d59-4756-9007-3454eac65880.gif)

FWIW, I did use zcash-cli with zcashd running in background to author a few transactions including one that moves some ZEC from the sapling pool to the orchard pool in a UA. That worked fine, and sync'ing continued after that. But it *might* be the first time I restarted zcashd after that when it started failing. No idea if that's related.

**Comment by steven-ecc** [comment]
@AArnott Does node resume normally after restarting? If not, can you try restarting with `-reindex`?

**Comment by AArnott** [comment]
No, the node won't start after several attempts. Or do you mean restarting my computer? 
I'm trying with `-reindex` at your suggestion and will let you know how that goes.

**Comment by AArnott** [comment]
It looked like it was about done with re-indexing, but now it appears it's re-_downloading_ the whole chain. I don't know why. It was sync'd to tip as of a day or two ago.

**Comment by AArnott** [comment]
It ultimately failed again, 26 hours later. 

```
     Downloading blocks | 1722335 (1722495 headers) / ~1725030 (99%)
                        | [                                                  ]
           Next upgrade | Unknown
            Connections | 8
  Network solution rate | 7.463 GSol/s

Since starting this node 1 days, 1 hours, 56 minutes, 4 seconds ago:
- You have validated 18438880 transactions!

Messages:
- Error: Error: A fatal internal error occurred, see debug.log for details

['zcash-cli.exe stop' to exit] [Set 'showmetrics=0' to hide]
```

The tail of debug.log is:

```
2022-07-03T19:57:39.445337Z  INFO main: UpdateTip: new best hash=000000000148db5a6c23f993b44a0bd5ef715397f3b5c1ae03437c304652b4f1 height=1722325 bits=469871101 log2_work=59.27322 tx=10935049 date=2022-07-01 11:38:46 progress=0.998342 cache=140.8MiB(61524tx)
2022-07-03T19:57:39.491913Z  INFO main: UpdateTip: new best hash=000000000019f23a16180be1793424d21b63516c2456c72a455708f581f8e24a height=1722326 bits=469871710 log2_work=59.273221 tx=10935051 date=2022-07-01 11:39:44 progress=0.998342 cache=140.8MiB(61525tx)
2022-07-03T19:57:39.522040Z  INFO main: UpdateTip: new best hash=000000000030ba142b6f02f9ca3cf0afcfaf9a59bbd8e923eb27de91bd50fb16 height=1722327 bits=469872240 log2_work=59.273222 tx=10935061 date=2022-07-01 11:43:38 progress=0.998344 cache=140.8MiB(61533tx)
2022-07-03T19:57:39.549692Z  INFO main: UpdateTip: new best hash=0000000000f36ea6a52af8a18879b7792d662bb4a5b1bf525f00c7e58b826bd7 height=1722328 bits=469871349 log2_work=59.273224 tx=10935063 date=2022-07-01 11:44:07 progress=0.998344 cache=140.8MiB(61533tx)
2022-07-03T19:57:39.569446Z  INFO main: UpdateTip: new best hash=0000000000a41e4ff94ca4d20447dd6839c05e22c814b1bd50a0860796aca280 height=1722329 bits=469870300 log2_work=59.273225 tx=10935064 date=2022-07-01 11:44:31 progress=0.998344 cache=140.8MiB(61534tx)
2022-07-03T19:57:39.598686Z  INFO main: UpdateTip: new best hash=000000000091e1edf19aa0fb3e60075c5458f78f2d792113ed9ecb0b67fa6785 height=1722330 bits=469876306 log2_work=59.273226 tx=10935066 date=2022-07-01 11:46:45 progress=0.998345 cache=140.8MiB(61536tx)
2022-07-03T19:57:39.616764Z  INFO main: UpdateTip: new best hash=00000000006e030c11b5efb3dc4c53986c786ec56f8378b18b9aab130b8c94a6 height=1722331 bits=469879503 log2_work=59.273228 tx=10935067 date=2022-07-01 11:47:13 progress=0.998346 cache=140.8MiB(61537tx)
2022-07-03T19:57:39.662122Z  INFO main: UpdateTip: new best hash=000000000187a16c268d82184175940dfbeb68a5049608cfe53f579f5bcb8331 height=1722332 bits=469879395 log2_work=59.273229 tx=10935070 date=2022-07-01 11:48:12 progress=0.998346 cache=140.8MiB(61539tx)
2022-07-03T19:57:39.682187Z  INFO main: UpdateTip: new best hash=0000000000a61f5c07d2c4daf9a7d7f1b34cc1c4b2f8a5e9212c138e023f252c height=1722333 bits=469879953 log2_work=59.27323 tx=10935071 date=2022-07-01 11:49:12 progress=0.998347 cache=140.8MiB(61540tx)
2022-07-03T19:57:39.704873Z  INFO main: UpdateTip: new best hash=00000000002aae8f44f06a259d793514070c0d147522bef412396e47ec3c95b5 height=1722334 bits=469879739 log2_work=59.273231 tx=10935075 date=2022-07-01 11:51:15 progress=0.998348 cache=140.8MiB(61549tx)
2022-07-03T19:57:39.832365Z  INFO main: UpdateTip: new best hash=0000000000477a558c2e9b72a50cd893440d9a70940d135066b17d50de8b151b height=1722335 bits=469880228 log2_work=59.273233 tx=10935083 date=2022-07-01 11:58:28 progress=0.998351 cache=140.8MiB(61556tx)
2022-07-03T19:57:41.710760Z ERROR main: ReadBlockFromDisk: OpenBlockFile failed for CBlockDiskPos(nFile=-1, nPos=0)
2022-07-03T19:57:41.711305Z  INFO main: *** ThreadNotifyWallets: Failed to read block 0000000000daa5abee13817b7becd1de139b5a2c5057f154fc02b74b1a3b3b3e while notifying wallets of block disconnects
2022-07-03T19:57:41.922222Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000010d62abdf9584769374c14806313bde9e3a632769f90f42903a66b7 height=1722336 bits=469883393 log2_work=59.273234 tx=10935087 date=2022-07-01 11:58:56 progress=0.998351 cache=140.8MiB(61560tx)
2022-07-03T19:57:42.020023Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=00000000016366384d0daf3d228ae8d73631b1c8591489547e6c7f05db96e1a6 height=1722337 bits=469880914 log2_work=59.273235 tx=10935089 date=2022-07-01 11:59:33 progress=0.998352 cache=140.8MiB(61558tx)
2022-07-03T19:57:42.100703Z  INFO main: tor: Thread interrupt
2022-07-03T19:57:42.101845Z  INFO main: torcontrol thread exit
2022-07-03T19:57:42.103246Z  INFO main: scheduler thread interrupt
2022-07-03T19:57:42.103252Z  INFO main: opencon thread interrupt
2022-07-03T19:57:42.103373Z  INFO main: addcon thread interrupt
2022-07-03T19:57:42.121145Z  INFO main: net thread interrupt
2022-07-03T19:57:42.151973Z  INFO ProcessNewBlock: main: UpdateTip: new best hash=0000000001a98706ac9a35a1c46f0b696d6424e0c01b5aeae1bda43a692c5463 height=1722338 bits=469882654 log2_work=59.273236 tx=10935094 date=2022-07-01 12:00:35 progress=0.998352 cache=140.8MiB(61566tx)
2022-07-03T19:57:42.152849Z  INFO main: msghand thread interrupt
2022-07-03T19:57:42.374739Z  INFO main: txnotify thread exit
2022-07-03T19:57:42.376551Z  INFO Shutdown: main: StopRPC: waiting for async rpc workers to stop
2022-07-03T19:57:42.963749Z  INFO Shutdown: main: StopNode()
2022-07-03T19:57:45.928374Z  INFO Shutdown: main: done
```

Any more ideas?

**Comment by str4d** [comment]
We believe this is fixed by #6054, which will be in release 5.1.0. If the issue persists after then, please re-open this issue.

**Comment by yangchou19** [comment]
I also meet the problems when I build in release 5.2.0

**Comment by ioptio** [comment]
I'm having this issue with 5.2.0. I did a reindex and it still happened. End of debug.log is a bunch of UpdateTip entries.

**Comment by ioptio** [comment]
Last entry before crash showed height 1759496 and when I restart zcashd the log shows 
`LoadBlockIndexDB: hashBestChain=000000000009c51a862245923ce123fa9558638aae44ea14d8e56fb67ad22bb3 height=1749563 date=2022-07-25 06:12:01 progress=0.977802`
and 
```
2022-08-27T10:50:56.509350Z  INFO Init: main: CWallet::InitLoadWallet(): Rescanning last 6384 blocks (from block 1743179)...
2022-08-27T10:50:56.509377Z  INFO Init: main: CWallet::ScanForWalletTransactions(): Rewinding Orchard wallet to height 1743178; current is 1759572
```
as it rescans

**Comment by ioptio** [comment]
After this last crash, it eventually synced up all the way. Not sure why it failed one more time after doing a reindex but so far it hasn't happened again.

**Comment by nuttycom** [comment]
This may be fixed by #6238, which is intended to be released in v5.3.1 - @daira please confirm.

**Comment by daira** [comment]
I believe this is fixed; the assertion messages are consistent with the issue that was fixed by #6238. Please file a new issue if you see a similar message again.
